x/superfluid/keeper/migrate.go

```diff
@@ -1,99 +1,177 @@
+		positionId, amount0, amount1, liquidity, joinTime, concentratedLockId, err = k.clk.CreateFullRangePositionUnlocking(ctx, concentratedPool, sender, exitCoins, remainingLockTime)
+		if err != nil {
+			return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, 0, 0, 0, err
+		}
+	}
+
+	// If the lock was previously superfluid delegated, superfluid delegate the new concentrated lock to the same validator
+	if wasSuperfluidDelegatedBeforeMigration {
+		err := k.SuperfluidDelegate(ctx, sender.String(), concentratedLockId, intermediateAccount.ValAddr)
+		if err != nil {
+			return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, 0, 0, 0, err
+		}
+	}
+
+	// If the lock was superfluid undelegating at time of migration
+	if wasSuperfluidUndelegatingBeforeMigration {
+		// Create and set a new intermediary account based on the previous validator but with the new lock id and concentratedLockupDenom
+		concentratedLockupDenom := cltypes.GetConcentratedLockupDenomFromPoolId(poolIdEntering)
+		clIntermediateAccount, err := k.GetOrCreateIntermediaryAccount(ctx, concentratedLockupDenom, valAddr)
+		if err != nil {
+			return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, 0, 0, 0, err
+		}
+
+		// Create a new synthetic lockup for the new intermediary account in an unlocking status
+		err = k.createSyntheticLockup(ctx, concentratedLockId, clIntermediateAccount, unlockingStatus)
+		if err != nil {
+			return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, 0, 0, 0, err
+		}
+	}
+
+	// If there are remaining gamm shares, we must re-lock them.
+	remainingGammShares := gammSharesInLock.Sub(sharesToMigrate)
+	if !remainingGammShares.IsZero() {
+		// Create a new lock with the remaining gamm shares for the remaining lock time.
+		newLock, err := k.lk.CreateLock(ctx, sender, sdk.NewCoins(remainingGammShares), remainingLockTime)
+		gammLockId = newLock.ID
+		if err != nil {
+			return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, 0, 0, 0, err
+		}
+
+		// If the gamm lock was previously superfluid bonded, superfluid delegate the gamm like normal
+		if wasSuperfluidDelegatedBeforeMigration {
+			err := k.SuperfluidDelegate(ctx, sender.String(), gammLockId, valAddr)
+			if err != nil {
+				return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, 0, 0, 0, err
+			}
+		}
+
+		// If the gamm lock was superfluid unbonding, get the previous gamm intermediary account, create a new gamm synthetic lockup, and set it to unlocking
+		if wasSuperfluidUndelegatingBeforeMigration {
+			gammIntermediateAccount, err := k.GetOrCreateIntermediaryAccount(ctx, remainingGammShares.Denom, valAddr)
+			if err != nil {
+				return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, 0, 0, 0, err
+			}
+			err = k.createSyntheticLockup(ctx, gammLockId, gammIntermediateAccount, unlockingStatus)
+			if err != nil {
+				return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, 0, 0, 0, err
+			}
+		}
```

Comment:
We check `if wasSuperfluidDelegatedBeforeMigration` or `if wasSuperfluidUndelegatingBeforeMigration` multiple times throughout the execution flow.

Let's decide on the strategy ones at the top of the function and then complete the full flow specific for that strategy.

This especially makes sense because undelegating and delegated are mutually exclusive so there is no need to have the same flow for the

Reply:
Agreed with the previous call of merging, but the remaining separation feels logical to me. Separated in three parts:

* Unlocking all previous locks
* Migrate what user desired
* Re-lock anything that user did not want to migrate